File: test/action-executor.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { spawnSync } from 'node:child_process';
import { mkdtempSync, rmSync, symlinkSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { createActionExecutor } from '../src/action-executor.js';
import { createConnector } from '../src/runtime/connector.js';
import { AgentNetworkHttpError } from '../src/runtime/httpClient.js';

const AUTHORIZATION_LIFETIME_MS = 60_000;

const MY_AGENT = { id: 'agent-me', name: 'My Agent', description: 'A useful Agent', tags: [] };
const TARGET = { id: 'agent-target', name: 'Research Agent' };
const REQUIREMENT = { id: 'requirement-1', status: 'draft', title: 'Research help' };

function memoryStore() {
  let value = null;
  return {
    async read() { return value; },
    async write(_name, next) { value = next; },
    async remove() {
      const existed = value !== null;
      value = null;
      return existed;
    },
    current() { return value; },
    set(next) { value = next; },
  };
}

function executorFor(responses) {
  const calls = [];
  const store = memoryStore();
  let index = 0;
  const api = {
    request: async (request) => {
      calls.push(request);
      const response = responses[index];
      index += 1;
      if (response instanceof Error) throw response;
      return response;
    },
  };
  return { calls, store, executor: createActionExecutor({ api, workflowStore: store, randomId: () => 'workflow-1' }) };
}

function authorizationHttp({ credentialStore, requests }) {
  return {
    async request(request) {
      requests.push(request);
      if (request.method === 'POST' && request.url.endsWith('/device-authorizations')) {
        return {
          authorizationId: request.body.authorizationId,
          userCode: 'ABCD-EFGH',
          verificationUrl: `https://tokensmind.ai/console/agent-network/authorize/${request.body.authorizationId}`,
          expiresAt: new Date(Date.now() + AUTHORIZATION_LIFETIME_MS).toISOString(),
          intervalSeconds: 1,
        };
      }
      const pending = credentialStore.current();
      if (request.url.includes('/device-authorizations/') && !request.url.endsWith('/exchange')) {
        return {
          authorizationId: pending.pending.authorizationId,
          expiresAt: pending.pending.authorization.expiresAt,
          status: 'approved',
        };
      }
      if (request.url.endsWith('/exchange')) {
        return {
          authorizationId: pending.pending.authorizationId,
          status: 'exchanged',
          credential: { id: 'credential-1', agentId: null, tokenPrefix: pending.pending.tokenPrefix },
        };
      }
      return [];
    },
  };
}

function credentialStoreAdapter(credentialStore) {
  return {
    async read(name) {
      const state = credentialStore.current() || {};
      return state[name] || null;
    },
    async write(name, value) {
      const state = credentialStore.current() || {};
      credentialStore.set({ ...state, [name]: value });
    },
    async remove(name) {
      const state = credentialStore.current() || {};
      const next = { ...state };
      delete next[name];
      credentialStore.set(next);
    },
  };
}

test('contact_agent performs the business loop behind one action', async () => {
  const { calls, executor, store } = executorFor([
    [MY_AGENT],
    [TARGET],
    REQUIREMENT,
    { requirement: { ...REQUIREMENT, status: 'open' }, recommendationCount: 1 },
    [{ agent: TARGET, canReceiveNewConversations: true }],
    { conversation: { id: 'conversation-1' }, created: true, message: { id: 1 } },
  ]);
  const result = await executor.execute({
    operation: 'contact_agent',
    input: {
      target: { name: TARGET.name },
      message: 'Let us discuss research collaboration.',
      requirement: { title: 'Research help', description: 'A sufficiently detailed research collaboration request.' },
    },
  });
  assert.equal(result.status, 'completed');
  assert.equal(result.data.created, true);
  assert.equal(calls.length, 6);
  assert.deepEqual(calls.map(({ method, path }) => `${method} ${path}`), [
    'GET /agent-network-api/agents?mine=1',
    'GET /agent-network-api/agents?q=Research%20Agent',
    'POST /agent-network-api/requirements',
    'POST /agent-network-api/requirements/requirement-1/publish',
    'GET /agent-network-api/requirements/requirement-1/recommendations',
    'POST /agent-network-api/conversations',
  ]);
  assert.equal(calls[2].idempotencyKey, 'workflow-1:requirement:create');
  assert.equal(calls[5].body.initialMessage.clientMessageId, 'workflow-1:conversation:create:message');
  assert.equal(store.current(), null);
});

test('contact_agent asks for missing input before making requests', async () => {
  const { calls, executor } = executorFor([]);
  const result = await executor.execute({ operation: 'contact_agent', input: { target: { id: TARGET.id } } });
  assert.equal(result.status, 'input_required');
  assert.deepEqual(calls, []);
});

test('search_agents requires a query and omits client paging controls', async () => {
  const missing = executorFor([]);
  const missingResult = await missing.executor.execute({
    operation: 'search_agents', input: { limit: 100 },
  });
  assert.equal(missingResult.status, 'input_required');
  assert.deepEqual(missingResult.fields, ['query']);
  assert.deepEqual(missing.calls, []);

  const found = executorFor([[TARGET]]);
  const result = await found.executor.execute({
    operation: 'search_agents', input: { query: 'Research Agent', limit: 100 },
  });
  assert.equal(result.status, 'completed');
  assert.equal(found.calls[0].path, '/agent-network-api/agents?q=Research%20Agent');
});

test('search_agents fails explicitly when the account has no Agent', async () => {
  const error = Object.assign(new Error('Agent profile required'), {
    code: 'AGENT_REQUIRED',
    status: 409,
  });
  const { executor } = executorFor([error]);
  const result = await executor.execute({
    operation: 'search_agents', input: { query: 'research' },
  });

  assert.equal(result.status, 'failed');
  assert.equal(result.code, 'AGENT_REQUIRED');
  assert.equal(result.details.nextOperation, 'ensure_agent');
  assert.match(result.message, /ensure_agent/);
});

test('ensure_agent sends optional public tags', async () => {
  const { calls, executor } = executorFor([[], { agent: MY_AGENT }]);
  const result = await executor.execute({
    operation: 'ensure_agent',
    input: { agent: MY_AGENT },
  });
  assert.equal(result.status, 'completed');
  assert.deepEqual(calls[1].body, {
    description: MY_AGENT.description,
    name: MY_AGENT.name,
    tags: MY_AGENT.tags,
  });
});

test('update_agent reads the owned profile then updates tags with a stable key', async () => {
  const updated = { ...MY_AGENT, tags: ['anime', 'community'] };
  const { calls, executor } = executorFor([[MY_AGENT], { agent: updated }]);
  const result = await executor.execute({
    operation: 'update_agent',
    input: { agent: { tags: updated.tags } },
  });
  assert.equal(result.status, 'completed');
  assert.equal(result.data.updated, true);
  assert.deepEqual(calls, [
    { method: 'GET', path: '/agent-network-api/agents?mine=1' },
    {
      body: { name: updated.name, description: updated.description, tags: updated.tags },
      idempotencyKey: 'workflow-1:agent:update',
      method: 'PATCH',
      path: '/agent-network-api/agents/agent-me',
    },
  ]);
});

test('ambiguous target is returned as selection_required', async () => {
  const { calls, executor } = executorFor([[MY_AGENT], [TARGET, { ...TARGET, id: 'agent-target-2' }]]);
  const result = await executor.execute({
    operation: 'contact_agent',
    input: {
      target: { name: TARGET.name },
      message: 'Let us discuss research collaboration.',
      requirement: { title: 'Research help', description: 'A sufficiently detailed research collaboration request.' },
    },
  });
  assert.equal(result.status, 'selection_required');
  assert.equal(result.candidates.length, 2);
  assert.equal(calls.length, 2);
});

test('failed retryable requests keep the action workflow', async () => {
  const { executor, store } = executorFor([Object.assign(new Error('temporary'), { status: 503, code: 'UPSTREAM_UNAVAILABLE' })]);
  const result = await executor.execute({ operation: 'get_my_agent', input: {} });
  assert.equal(result.status, 'failed');
  assert.equal(result.retryable, true);
  assert.equal(store.current().operation, 'get_my_agent');
});

test('invalid mark_read input never reaches the API', async () => {
  const { calls, executor } = executorFor([]);
  const result = await executor.execute({
    operation: 'mark_read',
    input: { conversationId: 'conversation-1', lastReadMessageId: 'not-a-number' },
  });
  assert.equal(result.status, 'input_required');
  assert.deepEqual(result.fields, ['lastReadMessageId']);
  assert.equal(calls.length, 0);
});

test('unblock_agent includes and encodes an optional blocker ID', async () => {
  const { calls, executor } = executorFor([{}]);
  const result = await executor.execute({
    operation: 'unblock_agent',
    input: { blockedAgentId: 'agent/blocked', blockerAgentId: 'agent/owner' },
  });
  assert.equal(result.status, 'completed');
  assert.equal(calls[0].path, '/agent-network-api/blocks/agent%2Fblocked?blockerAgentId=agent%2Fowner');
});

test('search without a credential authorizes and resumes the exact action', async () => {
  const credentialStore = memoryStore();
  const workflowStore = memoryStore();
  const requests = [];
  const opened = [];
  const http = authorizationHttp({ credentialStore, requests });
  const connector = createConnector({
    store: credentialStoreAdapter(credentialStore),
    browser: { open: async (url) => opened.push(url) },
    baseUrl: 'https://tokensmind.ai',
    http,
    client: { name: 'test', version: '1', deviceName: 'test', platform: 'test' },
  });
  const api = { request: (request) => connector.execute(request) };
  const executor = createActionExecutor({ api, workflowStore, randomId: () => 'workflow-1' });
  const request = { operation: 'search_agents', input: { query: 'research' } };

  const completed = await executor.execute(request);
  assert.deepEqual(completed, { status: 'completed', operation: 'search_agents', data: [] });
  assert.equal(opened.length, 1);
  assert.equal(requests.length, 4);
  assert.match(requests[3].url, /\/agent-network-api\/agents\?q=research$/);
  assert.match(requests[3].headers.Authorization, /^Bearer tm_agent_/);
  assert.equal(workflowStore.current(), null);
});

test('search HTTP 401 replaces the rejected credential and replays the search', async () => {
  const credentialStore = memoryStore();
  credentialStore.set({ active: { token: 'rejected-token', credentialId: 'old' } });
  const requests = [];
  const opened = [];
  const authorization = authorizationHttp({ credentialStore, requests });
  let rejected = false;
  const http = {
    async request(request) {
      if (!rejected && request.url.includes('/agents?q=research')) {
        rejected = true;
        requests.push(request);
        throw new AgentNetworkHttpError({
          status: 401, code: 'AUTH_REQUIRED', message: 'Sign in required',
        });
      }
      return authorization.request(request);
    },
  };
  const connector = createConnector({
    store: credentialStoreAdapter(credentialStore),
    browser: { open: async (url) => opened.push(url) },
    baseUrl: 'https://tokensmind.ai',
    http,
    client: { name: 'test', version: '1', deviceName: 'test', platform: 'test' },
  });

  const result = await connector.execute({
    method: 'GET', path: '/agent-network-api/agents?q=research',
  });

  assert.deepEqual(result, []);
  assert.equal(opened.length, 1);
  assert.equal(requests[0].headers.Authorization, 'Bearer rejected-token');
  assert.match(requests.at(-1).headers.Authorization, /^Bearer tm_agent_/);
  assert.notEqual(requests.at(-1).headers.Authorization, 'Bearer rejected-token');
});

test('search requires the stored credential and sends it to the service', async () => {
  const requests = [];
  const store = {
    async read(name) {
      if (name === 'active') return { token: 'stored-token', credentialId: 'credential-1' };
      return null;
    },
    async write() {},
    async remove() {},
  };
  const connector = createConnector({
    store,
    browser: { open: async () => {} },
    baseUrl: 'https://tokensmind.ai',
    http: { request: async (request) => { requests.push(request); return [TARGET]; } },
    client: { name: 'test', version: '1', deviceName: 'test', platform: 'test' },
  });

  const result = await connector.execute({
    method: 'GET',
    path: '/agent-network-api/agents?q=comics',
  });

  assert.deepEqual(result, [TARGET]);
  assert.equal(requests[0].headers.Authorization, 'Bearer stored-token');
});

test('Node CLI emits one structured result on stdout', () => {
  const script = path.join(path.dirname(fileURLToPath(import.meta.url)), '..', 'skills', 'tokensmind-agent-network', 'scripts', 'agent-network-runtime.mjs');
  const result = spawnSync(process.execPath, [script], {
    input: JSON.stringify({ operation: 'unsupported_operation', input: {} }),
    encoding: 'utf8',
  });
  assert.equal(result.status, 0);
  assert.deepEqual(JSON.parse(result.stdout), {
    status: 'input_required',
    message: 'Unsupported Agent Network operation: unsupported_operation',
    fields: ['operation'],
  });
  assert.equal(result.stderr, '');
});

test('Node CLI runs through an npm bin symlink', (context) => {
  const script = path.join(path.dirname(fileURLToPath(import.meta.url)), '..', 'skills', 'tokensmind-agent-network', 'scripts', 'agent-network-runtime.mjs');
  const tempDir = mkdtempSync(path.join(os.tmpdir(), 'agent-network-runtime-bin-'));
  const bin = path.join(tempDir, 'agent-network');
  context.after(() => rmSync(tempDir, { recursive: true, force: true }));
  symlinkSync(script, bin);
  const result = spawnSync(bin, [], {
    input: JSON.stringify({ operation: 'unsupported_operation', input: {} }),
    encoding: 'utf8',
  });
  assert.equal(result.status, 0);
  assert.equal(JSON.parse(result.stdout).status, 'input_required');
  assert.equal(result.stderr, '');
});

File: test_python/action_executor_test.py
import json
import os
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path

SCRIPT_DIR = Path(__file__).parents[1] / "skills" / "tokensmind-agent-network" / "scripts"
sys.path.insert(0, str(SCRIPT_DIR))

from agent_network_runtime import ActionExecutor  # noqa: E402
from python_runtime.connector import AgentNetworkConnector  # noqa: E402
from python_runtime.errors import AgentNetworkHttpError  # noqa: E402


class MemoryWorkflow:
    def __init__(self):
        self.value = None

    def read(self):
        return self.value

    def write(self, value):
        self.value = value

    def remove(self):
        existed = self.value is not None
        self.value = None
        return existed


class FakeApi:
    def __init__(self, responses):
        self.responses = list(responses)
        self.calls = []

    def request(self, method, path, body=None, *, key=None):
        self.calls.append({"method": method, "path": path, "body": body, "key": key})
        response = self.responses.pop(0)
        if isinstance(response, Exception):
            raise response
        return response


class AuthorizationStore:
    def __init__(self):
        self.values = {}

    def read(self, name):
        return self.values.get(name)

    def write(self, name, value):
        self.values[name] = value

    def remove(self, name):
        self.values.pop(name, None)


class AuthorizationHttp:
    def __init__(self, store, requests):
        self.store = store
        self.requests = requests

    def request(self, url, method="GET", headers=None, *, body=None):
        self.requests.append({"url": url, "method": method, "headers": headers, "body": body})
        if method == "POST" and url.endswith("/device-authorizations"):
            return {
                "authorizationId": body["authorizationId"],
                "userCode": "ABCD-EFGH",
                "verificationUrl": "https://tokensmind.ai/console/agent-network/authorize/%s" % body["authorizationId"],
                "expiresAt": "2099-01-01T00:00:00Z",
                "intervalSeconds": 1,
            }
        pending = self.store.read("pending")
        if url.endswith("/exchange"):
            return {
                "authorizationId": pending["authorizationId"],
                "status": "exchanged",
                "credential": {"id": "credential-1", "agentId": None, "tokenPrefix": pending["tokenPrefix"]},
            }
        if "/device-authorizations/" in url:
            return {"authorizationId": pending["authorizationId"], "expiresAt": pending["authorization"]["expiresAt"], "status": "approved"}
        return []


class ActionExecutorTest(unittest.TestCase):
    def test_contact_agent_performs_business_loop(self):
        mine = {"id": "agent-me", "name": "My Agent"}
        target = {"id": "agent-target", "name": "Research Agent"}
        requirement = {"id": "requirement-1", "status": "draft"}
        api = FakeApi([
            [mine], [target], requirement,
            {"requirement": {**requirement, "status": "open"}},
            [{"agent": target, "canReceiveNewConversations": True}],
            {"conversation": {"id": "conversation-1"}, "created": True, "message": {"id": 1}},
        ])
        executor = ActionExecutor(api, MemoryWorkflow())
        result = executor.execute({
            "operation": "contact_agent",
            "input": {
                "target": {"name": "Research Agent"},
                "message": "Let us discuss research collaboration.",
                "requirement": {
                    "title": "Research help",
                    "description": "A sufficiently detailed research collaboration request.",
                },
            },
        })
        self.assertEqual(result["status"], "completed")
        self.assertTrue(result["data"]["created"])
        self.assertEqual([call["method"] + " " + call["path"] for call in api.calls], [
            "GET /agent-network-api/agents?mine=1",
            "GET /agent-network-api/agents?q=Research%20Agent",
            "POST /agent-network-api/requirements",
            "POST /agent-network-api/requirements/requirement-1/publish",
            "GET /agent-network-api/requirements/requirement-1/recommendations",
            "POST /agent-network-api/conversations",
        ])
        self.assertTrue(api.calls[2]["key"].endswith(":requirement:create"))

    def test_missing_input_has_no_side_effect(self):
        api = FakeApi([])
        result = ActionExecutor(api, MemoryWorkflow()).execute({
            "operation": "contact_agent",
            "input": {"target": {"id": "agent-target"}},
        })
        self.assertEqual(result["status"], "input_required")
        self.assertEqual(api.calls, [])

    def test_ensure_agent_sends_optional_public_tags(self):
        profile = {
            "name": "My Agent",
            "description": "A useful research Agent.",
            "tags": ["research", "evidence"],
        }
        api = FakeApi([[], {"agent": {"id": "agent-me", **profile}}])
        result = ActionExecutor(api, MemoryWorkflow()).execute({
            "operation": "ensure_agent",
            "input": {"agent": profile},
        })
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[1]["body"], profile)

    def test_update_agent_reads_owned_profile_then_updates_it(self):
        current = {"id": "agent-me", "name": "My Agent", "description": "A useful Agent", "tags": []}
        updated = {**current, "tags": ["anime", "community"]}
        api = FakeApi([[current], {"agent": updated}])
        result = ActionExecutor(api, MemoryWorkflow()).execute({
            "operation": "update_agent",
            "input": {"agent": {"tags": updated["tags"]}},
        })
        self.assertEqual(result["status"], "completed")
        self.assertTrue(result["data"]["updated"])
        self.assertEqual(api.calls[1]["method"], "PATCH")
        self.assertEqual(api.calls[1]["path"], "/agent-network-api/agents/agent-me")
        self.assertEqual(api.calls[1]["body"]["tags"], updated["tags"])
        self.assertTrue(api.calls[1]["key"].endswith(":agent:update"))

    def test_unsupported_operation_is_input_required(self):
        result = ActionExecutor(FakeApi([]), MemoryWorkflow()).execute({
            "operation": "raw_request", "input": {},
        })
        self.assertEqual(result["status"], "input_required")

    def test_query_values_and_path_segments_are_encoded(self):
        api = FakeApi([[]])
        result = ActionExecutor(api, MemoryWorkflow()).execute({
            "operation": "search_agents",
            "input": {"query": "a/b? c"},
        })
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[0]["path"], "/agent-network-api/agents?q=a%2Fb%3F%20c")

    def test_search_agents_requires_query_and_omits_client_limit(self):
        missing_api = FakeApi([])
        missing = ActionExecutor(missing_api, MemoryWorkflow()).execute({
            "operation": "search_agents",
            "input": {"limit": 100},
        })
        self.assertEqual(missing["status"], "input_required")
        self.assertEqual(missing["fields"], ["query"])
        self.assertEqual(missing_api.calls, [])

        api = FakeApi([[]])
        result = ActionExecutor(api, MemoryWorkflow()).execute({
            "operation": "search_agents",
            "input": {"query": "research", "limit": 100},
        })
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[0]["path"], "/agent-network-api/agents?q=research")

    def test_search_agents_fails_explicitly_without_account_agent(self):
        error = AgentNetworkHttpError(409, "AGENT_REQUIRED", "Agent profile required")
        result = ActionExecutor(FakeApi([error]), MemoryWorkflow()).execute({
            "operation": "search_agents",
            "input": {"query": "research"},
        })
        self.assertEqual(result["status"], "failed")
        self.assertEqual(result["code"], "AGENT_REQUIRED")
        self.assertEqual(result["nextOperation"], "ensure_agent")
        self.assertIn("ensure_agent", result["message"])

    def test_unblock_agent_includes_encoded_blocker_id(self):
        api = FakeApi([{}])
        result = ActionExecutor(api, MemoryWorkflow()).execute({
            "operation": "unblock_agent",
            "input": {"blockedAgentId": "agent/blocked", "blockerAgentId": "agent/owner"},
        })
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[0]["path"], "/agent-network-api/blocks/agent%2Fblocked?blockerAgentId=agent%2Fowner")

    def test_http_failure_preserves_correction_and_retry_after(self):
        error = AgentNetworkHttpError(
            429,
            "RATE_LIMITED",
            "Too many requests",
            response={"correction": {"problem": "rate", "fix": "retry"}},
            headers={"retryAfter": "10"},
        )
        result = ActionExecutor(FakeApi([error]), MemoryWorkflow()).execute({
            "operation": "get_my_agent",
            "input": {},
        })
        self.assertEqual(result["status"], "failed")
        self.assertEqual(result["httpStatus"], 429)
        self.assertEqual(result["code"], "RATE_LIMITED")
        self.assertEqual(result["correction"]["fix"], "retry")
        self.assertEqual(result["retryAfter"], "10")

    def test_search_without_credential_authorizes_and_resumes(self):
        store = AuthorizationStore()
        requests = []
        connector = AgentNetworkConnector(
            store=store,
            browser=type("Browser", (), {"open": lambda self, url: None})(),
            base_url="https://tokensmind.ai",
            http=AuthorizationHttp(store, requests),
            client={"name": "test", "version": "1", "deviceName": "test", "platform": "test"},
        )
        operation = {"method": "GET", "path": "/agent-network-api/agents?q=research"}
        self.assertEqual(connector.execute(operation), [])
        self.assertEqual(len(requests), 4)
        self.assertTrue(requests[3]["url"].endswith("/agent-network-api/agents?q=research"))
        self.assertTrue(requests[3]["headers"]["Authorization"].startswith("Bearer tm_agent_"))

    def test_search_401_reauthorizes_and_replays(self):
        store = AuthorizationStore()
        store.write("active", {"token": "rejected-token", "credentialId": "old"})
        requests = []

        class ReauthorizationHttp(AuthorizationHttp):
            def __init__(self, authorization_store, captured):
                super().__init__(authorization_store, captured)
                self.rejected = False

            def request(self, url, method="GET", headers=None, *, body=None):
                if not self.rejected and "/agents?q=research" in url:
                    self.rejected = True
                    self.requests.append({"url": url, "method": method, "headers": headers, "body": body})
                    raise AgentNetworkHttpError(401, "AUTH_REQUIRED", "Sign in required")
                return super().request(url, method, headers, body=body)

        connector = AgentNetworkConnector(
            store=store,
            browser=type("Browser", (), {"open": lambda self, url: None})(),
            base_url="https://tokensmind.ai",
            http=ReauthorizationHttp(store, requests),
            client={"name": "test", "version": "1", "deviceName": "test", "platform": "test"},
        )

        result = connector.execute({"method": "GET", "path": "/agent-network-api/agents?q=research"})
        self.assertEqual(result, [])
        self.assertEqual(requests[0]["headers"]["Authorization"], "Bearer rejected-token")
        self.assertTrue(requests[-1]["headers"]["Authorization"].startswith("Bearer tm_agent_"))

    def test_python_cli_emits_one_structured_result(self):
        script = SCRIPT_DIR / "agent_network_runtime.py"
        with tempfile.TemporaryDirectory() as state_dir:
            environment = os.environ.copy()
            environment["TOKENSMIND_AGENT_NETWORK_STATE_DIR"] = state_dir
            result = subprocess.run(
                [sys.executable, str(script)],
                input=json.dumps({"operation": "unsupported_operation", "input": {}}),
                capture_output=True,
                text=True,
                check=False,
                env=environment,
            )
        self.assertEqual(result.returncode, 0)
        self.assertEqual(json.loads(result.stdout), {
            "status": "input_required",
            "message": "Provide a supported operation and object input.",
            "fields": ["operation", "input"],
        })
        self.assertEqual(result.stderr, "")


if __name__ == "__main__":
    unittest.main()
